reconcile/utils/terrascript_client.py

```diff
@@ -4073,17 +4093,86 @@ def populate_tf_resource_secrets_manager(self, resource, namespace_info):
 
         self.add_resources(account, tf_resources)
 
-    def populate_tf_resource_asg(self, resource, namespace_info):
+    def get_asg_image_id(self, image: Dict[str, Any],
+                         existing_secrets: Dict[str, Any], output_prefix: str,
+                         account: str, region: str) -> Tuple[str, str]:
+        """
+        AMI ID comes form queries result or AWS Api filter result.
+        AMI needs to be built by ci account and shared with target account.
+        AMI needs to be taged with a tag_name and
+        its value need to be the commit sha comes from upstream repo.
+        In case of there is a new commit while the packer job is still running,
+        use the AMI form existing asg.
+        """
+        image_id = image.get('id')
+        if image_id:
+            return image_id, ''
+
+        image_repo = image['repo']
+        tag_name = image_repo['tag_name']
+        url = image_repo['url']
+        ref = image_repo['ref']
+        pattern = r'^[0-9a-f]{40}$'
+        # get commit_sha from ref
+        if re.match(pattern, ref):
+            commit_sha = ref
+        # get commit_sha from branch
+        elif 'github' in url:
+            github = self.init_github()
+            repo_name = url.rstrip("/").replace('https://github.com/', '')
+            repo = github.get_repo(repo_name)
+            commit = repo.get_commit(sha=ref)
+            commit_sha = commit.sha
+        elif 'gitlab' in url:
+            raise ValueError("dose not support gitlab repo for now")
```

Comment:
can this be extracted to a function?
just thinking of the **future** refactor of this and saasherder (which has nearly the exact function _get_commit_sha).